[Draft] Polkadot Development Overview (build) (#4810)
w3f/polkadot-wiki: docs/build/build-guide.md

slug: ../build-guide
---

import Tabs from "@theme/Tabs";

import TabItem from "@theme/TabItem";

> Polkadot is a blockchain protocol with two goals: providing **shared security** among all connected

Reviewer: ```suggestion
{{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} is a blockchain protocol with two goals: providing **shared security** among all connected
```

parachains and allowing all connected chains to **interoperate** by using
[XCM](../learn/learn-xcm.md). With the advent of
[PDKs](../build/build-parachains.md#parachain-development-kit) like
[Substrate](https://substrate.io/) and [Cumulus](https://github.com/paritytech/cumulus), the time it
takes to develop and launch a new chain has dropped significantly. While before it would take years
to launch a new chain, now it may only take weeks or even days depending on your goals.

This guide will walk you through the steps you can take today to get started building your vision
with {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }}. It will explain the difference
between a [parachain](../learn/learn-parachains.md) and a smart contract in terms of deciding which
best fits your architectural needs.

This build guide provides four different tracks:

1. [Building Parachains](#building-parachains)
2. [Building a Pallet](#building-a-pallet)
3. [Developing Smart Contracts](#developing-smart-contracts)
4. [Developing a dApp](#developing-a-dapp)

:::info

> Keep in mind that these tracks are merely suggestive, and there are many ways to utilize Polkadot,

Reviewer: ```suggestion
Keep in mind that these tracks are merely suggestive, and there are many ways to utilize {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }},
```

Substrate, and their various developmental components. For more inspiration, take a look at the
[open source projects](./build-open-source.md) featured here in the wiki!

:::

> Before diving into any one of these tracks, it is encouraged to read about Polkadot and its networks

Reviewer: ```suggestion
Before diving into any one of these tracks, it is encouraged to read about {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} and its networks
```

in order to gain context about the application you could make.

## Development Ecosystem Overview

> ![Path to Building a Parachain](../assets/parachain-development.png)

Reviewer 2: I find this figure a bit convoluted. Especially the lines linking things to the cumulus box is unclear to me. Can you explain further?

Author: The idea was to sort of show the relationships between various components, and how one *could* morph from one to another.  Pallets make up your custom runtime -> Which are used by FRAME -> Which are both built on Substrate. 

The Cumulus part was to also show that really, they are all concluded as a runtime no matter the path you take (parachain or solo chain). 

 I'll work on making t clearer, probably a top-down approach is better 👍

Reviewer 2: Any updates on this?

Author: I will be working on it early next week - the education repo has been taking my time :) 

I'll request your review when I complete the last set of changes 👍

Author: Actually, it seems a lot of that repo may be relevant here!  I'll probably tag you in some of the diagrams I'm putting together over there :) Mostly to do with `frame` and their relationship with Substrate core libraries, that kind of stuff

Author: Working diagram - I think it's clearer than the other one so far.  Goal was to make it super clear what does what, and how they relate to one another

![Screenshot 2023-06-22 at 12 51 27 PM](https://github.com/w3f/polkadot-wiki/assets/26698074/f58a9514-6e69-4723-b97e-759065ba277f)

Author: If something like this is good, then I can theme it

edit: themed.

Reviewer 2: I am seeing this with a lot of delay, but the outcome is great, thanks!

application you want to make and understand the different paradigms available to developers who want
to build on {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }}.

## Building Parachains

![build 2](../assets/build-2.png)

> As mentioned, Polkadot is the **mainnet**, but is canonically referred to as the Relay Chain. It is

Reviewer: ```suggestion
{{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} is canonically referred to as the Relay Chain. It is
```

also considered a _layer zero_ protocol, as it enables for the interoperability and shared security
of multiple [parachains](../learn/learn-parachains.md), which are _layer one_ protocols. Parachains
connect to a relay chain using the [Parachains Protocol](../learn/learn-parachains-protocol.md).

![build 1](../assets/build-1.png)

:::info

Throughout this document, you will encounter the term **runtime** or **STF (State Transition
Function)**. Both refer to the same concept, as they both define how exactly a particular system,
i.e., a blockchain, should deal with state changes both externally and internally. Both of these

> terms are used extensively in Polkadot, and by association, Substrate contexts.

Reviewer: ```suggestion
terms are used extensively in {{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }}, and by association, Substrate contexts.
```

once every 24 hours (this makes the most sense if the data is useful to other players in the
{{ polkadot: Polkadot :polkadot }}{{ kusama: Kusama :kusama }} ecosystem too).

### Constructing FRAME Runtimes with Substrate

Polkadot is built using the [Substrate](https://substrate.io/) framework.
[Substrate](https://substrate.io/) is a highly configurable and dynamic framework that is used for
building blockchains. At a lower level, Substrate provides a set of tools and libraries ranging from
[block production, finality gadgets, to peer-to-peer networking](https://docs.substrate.io/reference/rust-api/).
Both Polkadot and Kusama, as well as most parachains are built using Substrate.

In essence, Substrate can break down a blockchain's development process by providing crucial
building blocks of functionality, removing the need for re-engineering complex mechanisms that
usually are involved when developing a blockchain.

Substrate can be used as a basis for a parachain to connect to a relay chain like Polkadot or
Kusama, or even as a basis to form a conventional layer one solo chain.

Currently, the most streamlined way of utilizing Substrate is

> [FRAME](https://docs.substrate.io/learn/runtime-development/#frame), which conveniently allows for a

Reviewer 2: In the future, we should try and gradually reduce links to `docs.substrate.io` and instead rely more and more on our RUST-docs in the short term, and in the long term (paritytech.github.io) on the new doc portal.

Author: That'd be awesome.  Right now, the only equivalent to those docs in the Rust docs is this one as far as I know 😅 https://paritytech.github.io/substrate/master/frame_support/macro.construct_runtime.html

Reviewer 2: Once https://github.com/paritytech/substrate/pull/14137/files is done, we can use it here :)

framework (based on Typescript) for crafting interactions with Substrate chains. It consists of a
development server and fluent API, which facilitates multichain interactions without compromising
either performance or ease of use. </TabItem> <TabItem value="subxt"> Submit extrinsics
(transactions) to a Substrate node via RPC using Rust. Also referred to as Rust Parity. </TabItem> </Tabs> -->

<Tabs groupId="clients" values={[ {label: 'Polkadot.js', value: 'pjs'}, {label: 'Subxt', value:
'subxt'}, {label: 'CAPI', value: 'capi'}, {label: 'React Hooks for ink!', value: 'ink'} ]}>

<TabItem value="pjs"> Promise and RxJS APIs around Polkadot and Substrate based chains via RPC
calls. It is dynamically generated based on what the Substrate runtime provides in terms of
metadata. Full documentation & examples available. </TabItem>

<TabItem value="capi"> Capi is a framework (based on Typescript) for crafting interactions with
Substrate chains. It consists of a development server and fluent API, which facilitates multichain
interactions without compromising either performance or ease of use. </TabItem>

> <TabItem value="subxt"> Submit extrinsics (transactions) to a Substrate node via RPC using Rust.

Reviewer 2: 👍
